torchao/sparsity/marlin.py

```diff
@@ -0,0 +1,443 @@
+
+    mask = _mask_creator(w.T).cuda().bool()
+    w = mask * w.T
+    w, meta = _sparse_semi_structured_from_dense_cutlass(w)
+    w = w.t()
+
+    in_features = in_features // 2
+
+    s = s.reshape((-1, out_features)).contiguous()
+    w = w.reshape((in_features // tile, tile, out_features // tile, tile))
+    w = w.permute((0, 2, 1, 3))
+    w = w.reshape((in_features // tile, out_features * tile))
+    res = w
+    res = res.reshape((-1, perm.numel()))[:, perm].reshape(res.shape)
+
+    # TODO(diogo): check if I can do this with torch?
+    q = np.zeros((res.shape[0], res.shape[1] // 8), dtype=np.uint32)
+    res = res.cpu().numpy().astype(np.uint32)
+    for i in range(8):
+        q |= res[:, i::8] << 4 * i
+
+    q = torch.from_numpy(q.astype(np.int32)).to(w.device)
+    q[:, :] = q.to(weight.device)
+    s[:, :] = s.to(scales.device)
+    meta[:, :] = meta.to(meta.device)
+
+    return q, s, meta
+
+
+def unpack_from_sparse_marlin_24(q: torch.Tensor, s: torch.Tensor, meta: torch.Tensor, n_tiles: int, initial_shape: torch.Size) -> Tuple[torch.Tensor, torch.Tensor]:
+
+    def unpack_scales(scales: torch.Tensor):
+        
+        # Step 1: Reverse the final reshape
+        # Original reshaped scales were of shape (-1, out_features)
+        s_unpacked = scales.reshape((-1, len(scale_perm)))
+
+        # Step 2: Reverse the permutation
+        reverse_perm = torch.tensor(scale_perm).argsort()
+        s_unpacked = s_unpacked[:, reverse_perm]
+
+        # Step 3: Reverse the initial reshape
+        # Flatten the scales back to 1D
+        s_unpacked = s_unpacked.reshape(-1)
+
+        return s_unpacked
+
+    def unpack_weights(q: torch.Tensor, tile: int, meta: torch.Tensor, initial_shape: Tuple[int, int]):
+        import numpy as np
+
+        res_ = np.zeros((q.shape[0], q.shape[1] * 8), dtype=np.uint32)
+        for i in range(8):
+            res_[:, i::8] = (q.cpu().numpy() >> (4 * i)) & 0xF
+        res_ = torch.from_numpy(res_.astype(np.int32)).to(q.device)
+
+        # Step 2: Reverse the permutation
+        res_ = res_.reshape((-1, perm.numel()))[:, torch.argsort(perm)].reshape(res_.shape)
+
+        # Step 3: Reverse the reshape and permutation applied before
+        in_features, out_features = initial_shape
+        in_features = in_features // 2
+
+        w = res_.reshape((in_features // tile, out_features // tile, tile, tile))
+        w = w.permute((0, 2, 1, 3))
+        w = w.reshape((in_features, out_features))
+        w = w.t()
+
+        w_unpacked = _sparse_semi_structured_to_dense_cutlass(w, meta)
+        w_unpacked = w_unpacked.t()
+
+        return w_unpacked
+
+    return unpack_weights(q, n_tiles, meta, initial_shape), unpack_scales(s)
+
+
+# This function converts dense matrix into sparse semi-structured
+# representation, producing "compressed" matrix, in the layout used by
+# CUTLASS backend, and corresponding metadata matrix.
+def _sparse_semi_structured_from_dense_cutlass(dense):
```

Comment:
I believe that this are the same functions that are available here: https://github.com/pytorch/pytorch/blob/main/torch/sparse/_semi_structured_conversions.py

So we don't have to copy them over.

Reply:
It seems some of them have small differences:
- https://www.diffchecker.com/ZWzCCGen/
- https://www.diffchecker.com/rIyyxGDx/
- https://www.diffchecker.com/kUzQSdpR/

Do you want me to leave it as is or should I change anything?

Reply:
cc @alexsamardzic I'm assuming they may have changed some stuff for int4 support. n00b question here, would it make sense to upstream some of the changes that they made?

@Diogo-V let's leave it for now, we can revisit once Alex responds.

Reply:
Yes - it would be good to have this upstream, but it would be good then also to have corresponding tests in `test/test_sparse_semi_structured.py` extended accordingly.

Reply:
I'll make a note; if taking on this, I'd also fully extend CUTLASS variant of SparseSemiStructuredTensor for int4 support.  Am at the moment working on stuff related to mixed data types MM, so can't promise exactly if/when I'll have cycles for this - sorry about that.